pkg/virt-handler/device-manager/device_controller.go

```diff
@@ -94,102 +163,72 @@ func (c *DeviceController) NodeHasDevice(devicePath string) bool {
-func (c *DeviceController) updatePermittedHostDevicePlugins() (map[string]ControlledDevice, map[string]ControlledDevice) {
-	devicePluginsToRun := make(map[string]ControlledDevice)
-	devicePluginsToStop := make(map[string]ControlledDevice)
+func (c *DeviceController) updatePermittedHostDevicePlugins() (map[string]controlledDevice, map[string]controlledDevice) {
+	devicePluginsToRun := make(map[string]controlledDevice)
+	devicePluginsToStop := make(map[string]controlledDevice)
 	// generate a map of currently started device plugins
-	for resourceName, hostDevDP := range c.devicePlugins {
-		_, isPermanent := permanentDevicePluginPaths[resourceName]
+	for resourceName, hostDevDP := range c.startedPlugins {
+		_, isPermanent := c.permanentPlugins[resourceName]
 		if !isPermanent {
 			devicePluginsToStop[resourceName] = hostDevDP
 		}
 	}
-	if hostDevs := c.virtConfig.GetPermittedHostDevices(); hostDevs != nil {
+	hostDevs := c.virtConfig.GetPermittedHostDevices()
+	if hostDevs == nil {
+		return devicePluginsToRun, devicePluginsToStop
+	}
+
+	if len(hostDevs.PciHostDevices) != 0 {
 		supportedPCIDeviceMap := make(map[string]string)
-		if len(hostDevs.PciHostDevices) != 0 {
-			for _, pciDev := range hostDevs.PciHostDevices {
-				log.Log.V(4).Infof("Permitted PCI device in the cluster, ID: %s, resourceName: %s, externalProvider: %t",
-					strings.ToLower(pciDev.PCIVendorSelector),
-					pciDev.ResourceName,
-					pciDev.ExternalResourceProvider)
-				// do not add a device plugin for this resource if it's being provided via an external device plugin
-				if !pciDev.ExternalResourceProvider {
-					supportedPCIDeviceMap[strings.ToLower(pciDev.PCIVendorSelector)] = pciDev.ResourceName
-				}
+		for _, pciDev := range hostDevs.PciHostDevices {
+			log.Log.V(4).Infof("Permitted PCI device in the cluster, ID: %s, resourceName: %s, externalProvider: %t",
+				strings.ToLower(pciDev.PCIVendorSelector),
+				pciDev.ResourceName,
+				pciDev.ExternalResourceProvider)
+			// do not add a device plugin for this resource if it's being provided via an external device plugin
+			if !pciDev.ExternalResourceProvider {
+				supportedPCIDeviceMap[strings.ToLower(pciDev.PCIVendorSelector)] = pciDev.ResourceName
 			}
-			pciHostDevicesPerResourceName := discoverPermittedHostPCIDevices(supportedPCIDeviceMap)
-			for pciResourceName, pciDevices := range pciHostDevicesPerResourceName {
-				log.Log.V(4).Infof("Discovered PCIs %d devices on the node for the resource: %s", len(pciDevices), pciResourceName)
-				// add a device plugin only for new devices
-				if _, isRunning := c.devicePlugins[pciResourceName]; !isRunning {
-					devicePluginsToRun[pciResourceName] = ControlledDevice{
-						devicePlugin: NewPCIDevicePlugin(pciDevices, pciResourceName),
-						stopChan:     make(chan struct{}),
-					}
-				} else {
-					delete(devicePluginsToStop, pciResourceName)
-				}
+		}
+		pciHostDevicesPerResourceName := discoverPermittedHostPCIDevices(supportedPCIDeviceMap)
```

Comment:
I also would like to point out that if you inline these, you don't need to come up with a name for the variable (I don't think you're re-using the variable anywhere):

```golang
for pciResourceName, pciDevices := range discoverPermittedHostPCIDevices(supportedPCIDeviceMap) {
...
}
```